Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -1492,101 +1493,202 @@ google_firestore_v1_MapValue Serializer::EncodeMapValue(
+
+  for (pb_size_t i = 0; i < size; i++) {
+    target_ids.push_back(array[i]);
+  }
+  return target_ids;
 }
 
 std::unique_ptr<WatchChange> Serializer::DecodeTargetChange(
     nanopb::Reader* reader,
     const google_firestore_v1_TargetChange& change) const {
-  // TODO(varconst): implement.
-  // - (std::unique_ptr<WatchChange>)decodedTargetChangeFromWatchChange:
-  //     (GCFSTargetChange *)change;
-  return {};
+  WatchTargetChangeState state =
+      DecodeTargetChangeState(reader, change.target_change_type);
+  auto target_ids =
+      DecodeTargetIdArray(reader, change.target_ids, change.target_ids_count);
+  ByteString resume_token = ByteString(change.resume_token);
+
+  util::Status cause;
+  if (change.has_cause) {
+    cause = util::Status{static_cast<Error>(change.cause.code),
+                         DecodeString(change.cause.message)};
+  }
+
+  return absl::make_unique<WatchTargetChange>(
+      state, std::move(target_ids), std::move(resume_token), std::move(cause));
 }
 
 WatchTargetChangeState Serializer::DecodeTargetChangeState(
     nanopb::Reader* reader,
     const google_firestore_v1_TargetChange_TargetChangeType state) {
-  // TODO(varconst): implement.
-  // -(WatchTargetChangeState)decodedWatchTargetChangeState:
-  //     (GCFSTargetChange_TargetChangeType)state;
-  return {};
+  switch (state) {
+    case google_firestore_v1_TargetChange_TargetChangeType_NO_CHANGE:
+      return WatchTargetChangeState::NoChange;
+    case google_firestore_v1_TargetChange_TargetChangeType_ADD:
+      return WatchTargetChangeState::Added;
+    case google_firestore_v1_TargetChange_TargetChangeType_REMOVE:
+      return WatchTargetChangeState::Removed;
+    case google_firestore_v1_TargetChange_TargetChangeType_CURRENT:
+      return WatchTargetChangeState::Current;
+    case google_firestore_v1_TargetChange_TargetChangeType_RESET:
+      return WatchTargetChangeState::Reset;
+    default:
+      HARD_FAIL("Unexpected TargetChange.state: %s", state);
+  }
 }
 
 std::unique_ptr<WatchChange> Serializer::DecodeDocumentChange(
     nanopb::Reader* reader,
     const google_firestore_v1_DocumentChange& change) const {
-  // TODO(varconst): implement.
-  // -(std::unique_ptr<WatchChange>)decodedDocumentChange:
-  //     (GCFSDocumentChange*)change;
-  return {};
+  ObjectValue value = ObjectValue::FromMap(DecodeFields(
+      reader, change.document.fields_count, change.document.fields));
+  DocumentKey key = DecodeKey(reader, change.document.name);
+  SnapshotVersion version =
+      DecodeSnapshotVersion(reader, change.document.update_time);
+  HARD_ASSERT(version != SnapshotVersion::None(),
+              "Got a document change with no snapshot version");
+  // The document may soon be re-serialized back to protos in order to store it
+  // in local persistence. Memoize the encoded form to avoid encoding it again.
+  Document document(std::move(value), key, version, DocumentState::kSynced,
+                    change.document);
```

Comment:
This is pretty dangerous.

The constructor of `Document` [takes its last argument](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/core/src/firebase/firestore/model/document.h#L69) as `absl::any`, which in many regards is similar to `void*`. Then the value of `any` will be unconditionally cast to an Objective-C type ([here](https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/Source/Local/FSTLocalSerializer.mm#L90)). `any` should throw if it's being cast to a different type than what it currently contains -- I'm surprised tests didn't catch this.

I'd suggest using the overload of the constructor without the last argument and adding a TODO to revisit this while porting the local serializer (where it's used).